src/ciphers/traits.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use generic_array::{typenum::Unsigned, ArrayLength, GenericArray};
+
+pub mod consts {
+    pub use generic_array::typenum::*;
+}
+
+pub type Key<T> = GenericArray<u8, T>;
+
+pub type Nonce<T> = GenericArray<u8, T>;
+
+pub type Tag<T> = GenericArray<u8, T>;
+
+pub trait Cipher {
+    /// The size of the key required by this algorithm.
+    type KeyLength: ArrayLength<u8>;
+
+    /// The size of the nonce required by this algorithm.
+    type NonceLength: ArrayLength<u8>;
+
+    /// The size of the tag produced by this algorithm.
+    type TagLength: ArrayLength<u8>;
+
+    type Error;
+
+    /// A human-friendly identifier of this algorithm.
+    const NAME: &'static str;
+
+    const KEY_LENGTH: usize = <Self::KeyLength as Unsigned>::USIZE;
+
+    const NONCE_LENGTH: usize = <Self::NonceLength as Unsigned>::USIZE;
+
+    const TAG_LENGTH: usize = <Self::TagLength as Unsigned>::USIZE;
+
+    fn padsize(_: &[u8]) -> usize {
+        0
+    }
+
+    fn encrypt(
+        key: &Key<Self::KeyLength>,
+        iv: &Nonce<Self::NonceLength>,
+        aad: &[u8],
+        ptx: &[u8],
+        ctx: &mut [u8],
+        tag: &mut Tag<Self::TagLength>,
```

Comment:
This is probably just my lack of cipher knowledge but will all these parameters always make sense for all kind of ciphers ?

Reply:
The very least, this should be called an AEAD trait. And use better names/descriptions for the parameters. Also nonce, and iv (initial vector) is not the same thing, so we should be consistent here.
Note that, `iv` and `aad` (which is suppose stands for associated data?) are optional, there are authenticating encryption schemes that do not rely on this.

Depending on the expected use-case, I would maybe propose to make this a GCM only trait and if necessary add a more general AE trait.

Reply:
Agreed that `Cipher` is not a good name but this is still an interface for AEAD, right? It was based on [`AeadInPlace`](https://docs.rs/aead/0.3.2/aead/trait.AeadInPlace.html) from the `aead` crate. I can go with the name `Aead`.

Will consistently use `nonce` and add more descriptive names 👍